packages/backend/src/misc/schema.ts

```diff
@@ -89,15 +115,39 @@ type NullOrUndefined<p extends Schema, T> =
 			? (T | undefined)
 			: T;
 
-export type SchemaType<p extends Schema> =
-	p['type'] extends 'number' ? NullOrUndefined<p, number> :
-	p['type'] extends 'string' ? NullOrUndefined<p, string> :
-	p['type'] extends 'boolean' ? NullOrUndefined<p, boolean> :
-	p['type'] extends 'array' ? NullOrUndefined<p, MyType<NonNullable<p['items']>>[]> :
+// 共用体型を交差型にする型 https://stackoverflow.com/questions/54938141/typescript-convert-union-to-intersection
+type UnionToIntersection<U> = (U extends any ? (k: U) => void : never) extends ((k: infer I) => void) ? I : never;
+
+export type UnionSchemaType<a extends ReadonlyArray<MinimumSchema>> =
+	a extends readonly [infer R, ...(infer X)] ?
+		MinimumSchemaType<R> | (X extends ReadonlyArray<MinimumSchema> ? UnionSchemaType<X> : never) :
+		never;
```

Comment:
元の定義を維持すると

```suggestion
export type UnionSchemaType<a extends ReadonlyArray<MinimumSchema>, X extends MinimumSchema =　a[number]> = X extends any ? UnionSchemaType<X> : never;
```

でよいはず（ほんまか）

Reply:
![image](https://user-images.githubusercontent.com/20679825/149615990-3758badc-28cc-48b9-89c7-21d4898f15dc.png)

Reply:
なんか…気色悪い

Reply:
`T` が type alias であるとき `T extends something ? (ifTrue) : (ifFalse)` における `ifTrue` および `ifFalse` 内の `T` は union が分解される挙動があるので、だいたい `T extends any` と書かれていたらこの挙動を利用したいという意図があるものとして読んで良いと思う
![image](https://user-images.githubusercontent.com/20679825/149616539-9ccd819d-67ca-41c9-915a-0f2e503661b2.png)

Reply:
type alias そのものにしか効かないので、当然 `a[number]` と書いていてそのまま `a[number] extends any` としてもこの挙動は発生しない